File: packages/registry/src/define.ts
import semver from "semver";
import { z } from "zod";
import type { ComponentDefinition } from "./types.js";

export class ComponentDefinitionError extends Error {
  constructor(type: string, message: string) {
    super(`component "${type}": ${message}`);
    this.name = "ComponentDefinitionError";
  }
}

const TYPE_RE = /^[a-zA-Z][a-zA-Z0-9_]*(\.[a-zA-Z][a-zA-Z0-9_]*)*$/;

/**
 * Validates and freezes a component definition.
 * Attempts the JSON Schema conversion at definition time and fail-fast rejects JSON-unrepresentable
 * props types (z.date, etc.) back to the catalog author.
 */
export function defineComponent<P extends z.ZodObject>(def: ComponentDefinition<P>): ComponentDefinition<P> {
  if (!TYPE_RE.test(def.type)) {
    throw new ComponentDefinitionError(def.type, "type must match dotted identifier form");
  }
  if (semver.valid(def.version) == null) {
    throw new ComponentDefinitionError(def.type, `version "${def.version}" is not valid semver`);
  }
  if (def.description.trim().length === 0) {
    throw new ComponentDefinitionError(def.type, "description is required (used in LLM prompt)");
  }
  try {
    z.toJSONSchema(def.propsSchema, {
      target: "draft-2020-12",
      unrepresentable: "throw",
      reused: "inline",
    });
  } catch (e) {
    throw new ComponentDefinitionError(
      def.type,
      `propsSchema is not JSON-representable: ${e instanceof Error ? e.message : String(e)}`,
    );
  }
  // Default implementation to native when unspecified (stop depending on spread order and make the intent explicit).
  return Object.freeze({ ...def, implementation: def.implementation ?? { kind: "native" as const } });
}

File: python/kohaku/src/kohaku/registry/define.py
"""Component-definition validation and freezing (Port of TS define.ts)."""

from __future__ import annotations

import re

from kohaku.spec import canonical_stringify

from .semver_util import is_valid
from .types import ComponentDefinition


class ComponentDefinitionError(Exception):
    def __init__(self, type_: str, message: str) -> None:
        super().__init__(f'component "{type_}": {message}')


_TYPE_RE = re.compile(r"^[a-zA-Z][a-zA-Z0-9_]*(\.[a-zA-Z][a-zA-Z0-9_]*)*$")


def define_component(definition: ComponentDefinition) -> ComponentDefinition:
    """At definition time, fail-fast any inconsistency back to the catalog author.

    The TS version checks the representability of the zod -> JSON Schema conversion, but the Python version's
    propsSchema is already a JSON Schema, so it checks "whether it can be canonicalized as a JSON value"
    (rejecting functions or non-JSON values here).
    """
    if not _TYPE_RE.match(definition.type):
        raise ComponentDefinitionError(definition.type, "type must match dotted identifier form")
    if not is_valid(definition.version):
        raise ComponentDefinitionError(
            definition.type, f'version "{definition.version}" is not valid semver'
        )
    if len(definition.description.strip()) == 0:
        raise ComponentDefinitionError(
            definition.type, "description is required (used in LLM prompt)"
        )
    try:
        canonical_stringify(definition.propsSchema.json_schema)
    except TypeError as e:
        raise ComponentDefinitionError(
            definition.type, f"propsSchema is not JSON-representable: {e}"
        ) from None
    return definition
